mmv1/templates/terraform/constants/service_networking_vpc_service_controls.go.erb

```diff
@@ -0,0 +1,76 @@
+func resourceServiceNetworkingVPCServiceControlsSet(d *schema.ResourceData, meta interface{}, config *transport_tpg.Config) error {
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	project, err := tpgresource.GetProject(d, config)
+	if err != nil {
+		return err
+	}
+	projectNumber, err := getProjectNumber(d, config, project, userAgent)
+	if err != nil {
+		return err
+	}
+
+	network := d.Get("network").(string)
+	enabled := d.Get("enabled").(bool)
+
+	obj := make(map[string]interface{})
+	obj["consumerNetwork"] = fmt.Sprintf("projects/%s/global/networks/%s", projectNumber, network)
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{ServiceNetworkingBasePath}}services/{{service}}")
+	if err != nil {
+		return err
+	}
+
+	if enabled {
+		url = url + ":enableVpcServiceControls"
```

Comment:
Precreate + Preread custom code is pretty atypical, and I want to make sure I'm understanding the reasoning behind their inclusion. Is it because of this enabled vs disabled logic?

Reply:
The preread is to translate the project ID into a project number as required by the service networking API. This translation is needed to build the correct URL. Note that this same issue has affected other [similar resources](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/servicenetworking/resource_google_service_networking_peered_dns_domain.go#L243)

The custom_create/custom_update is needed, again, because of the atypical URLs used by this API. The API uses distinct URLs for [enabling](https://cloud.google.com/service-infrastructure/docs/service-networking/reference/rest/v1/services/enableVpcServiceControls) and [disabling](https://cloud.google.com/service-infrastructure/docs/service-networking/reference/rest/v1/services/disableVpcServiceControls) the controls, and doesn't involve creating or updating a traditional resource. Since the URL depends on the desired state, we need custom logic to construct the correct request.

Reply:
Thanks for the explanation and additional context.